server/sandbox_run.go

```diff
@@ -80,7 +83,93 @@ func (s *Server) trustedSandbox(req *pb.RunPodSandboxRequest) bool {
 }
 
 func (s *Server) runContainer(container *oci.Container, cgroupParent string) error {
-	if err := s.Runtime().CreateContainer(container, cgroupParent); err != nil {
+	intermediateMountPoint := container.IntermediateMountPoint()
+
+	if intermediateMountPoint == "" {
+		if err := s.Runtime().CreateContainer(container, cgroupParent); err != nil {
+			return err
+		}
+	} else {
+		errc := make(chan error)
+		go func() {
+			// We create a new mount namespace before running the container as the rootfs of the
+			// container is accessible only to the root user.  We use the intermediate mount
+			// namespace to bind mount the root to a directory that is accessible to the user which
+			// maps to root inside of the container/
+			var err error
+			runtime.LockOSThread()
+			defer func() {
+				if err == nil {
+					runtime.UnlockOSThread()
```

Comment:
shouldn't you unlock regardless of the err condition?

Reply:
that wouldn't be safe if we fail at restoring the previous namespace.  If we unlock regardless of that, the thread might continue its execution in the intermediate mount namespace we are setting up here.

Reply:
@giuseppe I think it would be good to add this as a comment in the code.